### ISSUE TITLE
POC to read JIRA issues by filter and custom fields

### ISSUE THREAD

**Issue by vorburger** [vorburger/opendaylight-bot#7]
The goal of this issue is to raise a PR with a POC that introduces a class which:

* given as input (1) a JIRA base URL such as `https://jira.opendaylight.org` and (2) a JIRA QL (?) filter such as `project = tsc AND issuetype = "Weather Item" AND status not in (closed, Resolved)` returns a List<String> of issues matching that filter - currently that would be TSC-99, TSC-101 and TSC-102.

* given as a input (1) a JIRA base URL and (2) the ID of a JIRA issue and (3) a list of custom field name/IDs, returns the values of those fields. Specifically, given e.g. TSC-99 for https://jira.opendaylight.org/browse/TSC-99 and "ODL Managed Gerrit Topic:" and "ODL Last FULL Build:" (or it's numeric ID, if you can find it) it would return "binding-tlc-rpc" and https://jenkins.opendaylight.org/...

**Comment by ivange94** [comment]
@vorburger can you assign me to this?

**Comment by ivange94** [comment]
@vorburger when I launch the bot, I see this interface 
<img width="784" alt="screen shot 2018-06-03 at 7 56 23 am" src="https://user-images.githubusercontent.com/11330494/40884064-b2bb21b2-6703-11e8-9dc2-5710315cda91.png">

I'm to add fields that will receive the input above or how will this input be sent to the bot?

**Comment by ivange94** [comment]
@vorburger I intend to use the Jira REST API here. I was wondering which version of jira is at https://jira.opendaylight.org. I mean is, it jira cloud or server? They seem to have separate documentations.

**Comment by vorburger** [comment]
> @vorburger can you assign me to this?

I don't seem to be able to - not sure how GitHub determines the elibible list of assignees.

Doesn't matter - let's just say that until you say so otherwise by comment, this issue is "yours".

> I'm to add fields that will receive the input above or how will this input be sent to the bot?

no fields needed; let's the goal is to show be a table (can be ugly, just <ul><li> is fine) on that homepage of the bot with the JIRA IDs resulting from running a query matching the filter shown above.

A JUnit for a new `class JIRA` with a method like List<Issue> getIssues(String filter) would be nice.

> which version of jira is at https://jira.opendaylight.org. I mean is, it jira cloud or server? 

pretty sure that is JIRA v7.6.0 Server run by LF, not Cloud hosted by Atlassian, did you see the footer?

https://developer.atlassian.com/server/jira/platform/java-apis/ looks interesting BTW.

**Comment by ivange94** [comment]
> no fields needed; let's the goal is to show be a table (can be ugly, just `<ul><li>` is fine) on that homepage of the bot with the JIRA IDs resulting from running a query matching the filter shown above.

Ok


> A JUnit for a new class JIRA with a method like List getIssues(String filter) would be nice.

You mean a unit test?

> https://developer.atlassian.com/server/jira/platform/java-apis/ looks interesting BTW.

Was planing to use the REST API but this looks much more cooler

**Comment by ivange94** [comment]
Working on something now. Should have a PR soon.

**Comment by ivange94** [comment]
```xml
<dependency>
      <groupId>com.atlassian.jira</groupId>
      <artifactId>jira-api</artifactId>
      <version>7.6.1</version>
      <scope>provided</scope>
    </dependency>
```

I'm having trouble resolving this dependency. InteliJ can't find it. **mvn clean install** from the terminal in InteliJ gives

> [ERROR] Failed to execute goal on project bot: Could not resolve dependencies for project org.opendaylight.bot:bot:jar:1.0.0-SNAPSHOT: Failure to find com.atlassian.jira:jira-api:jar:7.6.1 in https://nexus.opendaylight.org/content/repositories/public/ was cached in the local repository, resolution will not be reattempted until the update interval of opendaylight-mirror has elapsed or updates are forced -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
> [ERROR] 
> [ERROR] After correcting the problems, you can resume the build with the command 

When I force the update with the -U switch, I get this

> INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal on project bot: Could not resolve dependencies for project org.opendaylight.bot:bot:jar:1.0.0-SNAPSHOT: Could not find artifact com.atlassian.jira:jira-api:jar:7.6.1 in opendaylight-mirror (https://nexus.opendaylight.org/content/repositories/public/) -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
> [ERROR] 
> [ERROR] After correcting the problems, you can resume the build with the command

Surprisingly the build passes when I run **mvn clean install** from my normal terminal(Not the one in IDEA). But the dependency is still not downloaded.

**ls ~/.m2/repository/com/atlassian/jira/jira-api/** tells me directory doesn't exist.


For now I'll proceed with a REST client and talk to Jira via REST. Will come back to this later.

**Comment by ivange94** [comment]
I had to add this https://ecosystem.atlassian.net/wiki/spaces/JRJC/pages/27164674/Download to my settings.xml for that error to disappear but I'm getting a new error. Duplicate classes found. See https://pastebin.com/tNsz33UN.

I'm leaving for school now. Will get back to this when I return.

**Comment by prateekiiest** [comment]
Hi @ivange94 .  I would request you to check the pom.xml and settings.xml for any duplicate dependency (like adding the same dependency twice), in case if you have mistakenly done so.

Also have you checked if  `jira-api` dependency is present on `http:/repo/maven` ?
You can check this issue https://github.com/vorburger/opendaylight-bot/issues/20 for similar issue I faced earlier

I hope @vorburger would be able to add further on this :)

**Comment by vorburger** [comment]
@ivange94 yeah this is a similar problem to that @prateekiiest faced on #20; basically:

So `com.atlassian.jira:jira-api` is available (only) on [maven.atlassian.com](https://maven.atlassian.com/public/com/atlassian/jira/) but [not on Maven Central](http://search.maven.org/#search%7Cga%7C1%7Cjira-api).  

So you can add `maven.atlassian.com` to your settings.xml, but that is always a PITA... you locally do that, but it can confuse future contributors. And if this code ever moves to http://git.opendaylight.org, then it's a PITA again - because https://nexus.opendaylight.org does not mirror `maven.atlassian.com`, and requesting that may be possible but woud likely be a bigger discussion - premature, for this project.

What you can try however is to add `maven.atlassian.com` not to a settings.xml but into `pom.xml`, [see here on maven.apache.org for how to do that](http://maven.apache.org/guides/mini/guide-multiple-repositories.html). That approach has other disadvantages, but I think to unblock us here, it's probably a fair short term solution.

Re. `duplicate-finder-maven-plugin`, that's an entirely different issue. You can probably fix that by [adding an exclusion like I have here](https://github.com/vorburger/opendaylight-bot/commit/822fd444797c68a0c3606eee5c9e781cdf906f9e) for `stax-api` into your `<dependency>` for `jira-api`, try it. (It's because `stax-api` overlaps with and contains the same classes than `xml-apis`, which is very bad, but `xml-apis` contains those and more, I just checked, so the right thing to do is exclude `stax-api`.)
